### ISSUE TITLE
Failure to format `struct Array`

### ISSUE THREAD

**Issue by MaxDesiatov** [nicklockwood/SwiftFormat#1144]
Repro code:

```swift
@frozen
public struct Array<Element> {
}
```

When trying to format this snippet, SwiftFormat prints an error message: `error: Expected identifier on line 2 in Array.swift.`

**Comment by nicklockwood** [comment]
@MaxDesiatov I'm not able to repro this. What SwiftFormat version and config are you using?

**Comment by nicklockwood** [comment]
@MaxDesiatov or is there perhaps some code before this line that is needed to trigger the bug?

**Comment by MaxDesiatov** [comment]
I'm using 0.49.4 from Homebrew, it's reproducible for me with no config and also with either this config:

```
--indent 2
--indentcase false
--trimwhitespace always
--voidtype tuple
--nospaceoperators ..<,...
--ifdef noindent
--stripunusedargs closure-only
--maxwidth 100
--wraparguments before-first
--funcattributes prev-line
--disable andOperator
--swiftversion 5.4
```

or this one:

```
--disable redundantReturn
--maxwidth 80
--closingparen same-line
--emptybraces linebreak
--disable wrapMultilineStatementBraces
--redundanttype explicit
--wraparguments after-first
--extensionacl on-declarations
```

> or is there perhaps some code before this line that is needed to trigger the bug?

No, those three lines (and a trailing newline) is the sole content of the file.

**Comment by MaxDesiatov** [comment]
Also reproducible with the VSCode SwiftFormat extension:

<img width="554" alt="Screenshot 2022-02-21 at 13 13 55" src="https://user-images.githubusercontent.com/112310/154962302-27ebb737-3538-4899-a86e-6907a1f15a6b.png">

**Comment by nicklockwood** [comment]
@MaxDesiatov OK, I can repro now. For some reason it didn't happen when I had the code in a playground.

**Comment by nicklockwood** [comment]
@MaxDesiatov this is caused by the `typeSugar` rule trying to convert the `Array<Element>` to `[Element]`. It's a pretty unusual edge case since people aren't normally declaring their own `Array` implementation. I'll add a fix, but you're going to want to disable that rule fro your project anyway because it will break a lot of other things.

**Comment by MaxDesiatov** [comment]
To clarify, we're developing [an alternative for Swift stdlib](https://github.com/compnerd/uswift) for bare metal and other use cases, where keeping binaries small is critical. It does require reimplementing stdlib types from scratch, but I found SwiftFormat is quite close to coding standards established in the library already.

> this is caused by the `typeSugar` rule trying to convert the `Array<Element>` to `[Element]`. It's a pretty unusual edge case since people aren't normally declaring their own `Array` implementation. I'll add a fix, but you're going to want to disable that rule for your project anyway because it will break a lot of other things.

What kind of breakage should we expect? This rule only reformats `Array<Int>` as `[Int]`, right?

**Comment by nicklockwood** [comment]
> What kind of breakage should we expect? This rule only reformats Array<Int> as [Int], right?

Right. The issue is that the Swift compiler won't recognize your overloaded `Array<Element>` as being the same as the one in the standard library, so in this case `[Int]` and `Array<Int>` don't refer to the same type:

```swift
struct Array<Element> {}

let foo: [Int] // refers to stdlib Array, not local Array
```

I'm not sure if that's actually a problem in your case though, if you are writing your own stdlib replacement then it might just work. I'll add the fix anyway, but in the meantime you can either disable the rule, or just do this if you'd prefer to keep it enabled generally:

```swift
@frozen
// swiftformat:disable:next typeSugar
public struct Array<Element> {
}
```

**Comment by MaxDesiatov** [comment]
Thanks, this makes sense.

> the Swift compiler won't recognize your overloaded `Array<Element>` as being the same as the one in the standard library, so in this case `[Int]` and `Array<Int>` don't refer to the same type:

This isn't the case, since uSwift isn't built as an overlay or an addition to stdlib, but as a complete replacement. It isn't supposed to be used in a normal environment, where stdlib is already available. By passing `-parse-stdlib` flag to `swiftc` we make it "forget" about already installed stdlib. Our `Array` isn't override, the compiler treats it as the real deal.

**Comment by nicklockwood** [comment]
@MaxDesiatov fixed in 0.49.5
